---
title: 'LAs BEST: Genome Analysis'
author: "David Conti"
date: "`r format(Sys.time(), '%d %B %Y')`"

output:
  html_document:
    toc: true
    toc_float: true
    df_print: paged
    code_folding: hide
---

```{r genome setup, include=FALSE, echo=FALSE}
library(knitr)
library(ggplot2)
library(corrplot)
library(RColorBrewer)
library(here)

options(knitr.table.format = "html")
knitr::opts_chunk$set(echo = TRUE)

# folder for analysis
#setwd(here::here("LA.Best/2022"))

# suppress warning messages for final rendering
old.warn <- getOption("warn")
options(warn=-1)

```

# ALLELE FREQUENCY ESTIMATION
## Specify the true values for the population
```{r input parameters }
# input parameters
True_Freq_p <- 0.35
True_Freq_q <- 1 - True_Freq_p
N <- 1000
```

## Simulate a genotype for a population
How can we use R to create a genotype for all individuals that follows a binomial distribution?
```{r simulate a genotype, echo=TRUE }
# binomial sampling
G <- rbinom(N, 2, True_Freq_p)  # binomial sampling

```

## Estimate Allele Frequencies
Given a vector of genotypes for all individuals, how can we estimate allele frequencies using the counts of each genotype?
```{r estimate allele frequencies, echo=TRUE }
#create table of genotype counts
n <- table(G)

# estimate allele frequencies via equation on slide 16
p <- (2*n[3] + n[2])/(2*N)
q <- (2*n[1] + n[2])/(2*N)
cat("True frequency: ", True_Freq_p, "Estimated frequency: ", p)

```

## Estimate Allele Frequencies using MLE {.tabset}
Given a vector of genotypes for all individuals, how can we estimate allele frequencies by maximizing the log-likelihood?

### Write out the log-likelihood for a binomial:
```{r log-likelihood binomial, echo=TRUE }
# log Likelihood
logLik_function <- function(v_p, G) {
  logLik_i <- G*log(v_p) + (2-G)*log(1-v_p)
  logLik <- sum(logLik_i)
}
```

### Visualize maximization of the log-likelihood across a range of allele frequencies
Plot the values of the log-likelihood for each specified value of the allele frequencies
```{r visualize MLE, echo=TRUE }
# demonstrate the values of the log likelihood for a range of frequencies
p.range.values <- seq(from=0.01, to=0.99, by=.01)
logLik_values <- sapply(p.range.values, logLik_function, simplify=T, G=G)
plot(p.range.values, logLik_values, xlab="Allele Frequency (p)", ylab="Likelihood Value", type="l", lwd=2)
abline(h=max(logLik_values), lwd=2, col="red")
abline(v=True_Freq_p, col="blue", lwd=2)
```

### Find the MLE using an opitmization algorithm
```{r maximize MLE}
# using 'otim' function to estimate allele frequencies
MLE <-  optim(par=c(0.5), # initial values for frequency
              fn = logLik_function, # function to maximize
              method = "L-BFGS-B", # allows box constraints each variable can be given a lower and/or upper bound
              lower=c(0.00001),
              upper=c(.999999),
              control = list(fnscale = -1), # maximize the function
              hessian = T, # calculate Hessian matrix because we will need for confidence intervals
              G=G  # data
)

# output parameter estimate:
cat("True frequency: ", True_Freq_p, "Estimated frequency with MLE: ", MLE$par)

```


# DISEASE ASSOCIATION
## Simulate a disease from the genotype for a population
How can we use R to create a disease phenotype for all individuals that is the Pr(Disease=1 | Genotype)?

### Input parameters
```{r penetrance parameters for disease, echo=TRUE }
# Pr(Disease = 1 | Genotype)
PrD_G_0 <- 0.1
PrD_G_1 <- 0.2
PrD_G_2 <- 0.3

```

### Simulate Disease Phenotype
```{r simulate a disease, echo=TRUE }
#create vector of probabilities for all individuals
PrD_G <- ifelse(G==0, PrD_G_0, ifelse(G==1, PrD_G_1, PrD_G_2))
# binomial sampling with conditions
Y <- rbinom(N, 1, PrD_G)  # binomial sampling
```

### Perform Chi-Squared Test
```{r test association for disease, echo=TRUE }
r.chisqtest <- chisq.test(G, Y)
print(r.chisqtest)
```


# GENOME_WIDE ASSOCIATION STUDY
## GWAS Data Input
```{r Genome: Data Analysis setup, echo=TRUE }
load("genome.RData")

# Other variables for analysis
N <- nrow(X) # number of individuals in the analysis
Q <- ncol(U)  # number of covariates in the matrix U
P <- ncol(X)  # number of SNPs in the matrix X

```


## Genome: Univariate Regression
```{r Genome: Univariate model, echo=TRUE}
snp.Names <- paste("SNP", 1:P, sep="_")
univariate.results <- t(sapply(1:P, FUN=function(p) {  # using index p facilitate writing results
  x <- X[,p]
  reg <- glm(Y~x+U, family=binomial)    # perform logistic regression
  s.reg <- summary(reg)                 # get the summary for the regression
  c.reg <- s.reg$coef["x",]             # select the coefficients for the exposure
  write.table(t(c(snp.Names[p], c.reg)), file="GenomeUnivariateResults.txt", append=ifelse(p==1, F, T), quote=F, sep="\t", col.names=ifelse(p==1, T, F), row.names=F)
  return(c.reg)                         # to avoid potential memory issues only return coefficients if small number of exposures
}, simplify=T))
univariate.results <- data.frame(snp.Names,univariate.results)
names(univariate.results) <- c("SNP.Name","Estimate", "SD","Z.statistic", "P.value")
univariate.results$P.value <- format(univariate.results$P.value, scientific=T)


```

## Univariate results: {.tabset}
### Univariate Summary Table:
```{r Genome: Univariate table}
kable(univariate.results[as.numeric(univariate.results$P.value)<0.05,], digits=3, align="c", row.names=FALSE, col.names=c("SNP","Estimate", "SD","Z Statistics", "P-value"))
```

### Univariate Manhattan Plot:
```{r Genome: Univariate MH plot}
neglog.pvalues <- -log10(as.numeric(univariate.results$P.value))
plot(1:nrow(univariate.results), neglog.pvalues, 
     pch=16, xaxt="n", ylim=c(0, max(neglog.pvalues, 3)),
     ylab="-log(p-value)", xlab="SNPs")
abline(h=-log10(0.05/nrow(univariate.results)), lty=2, lwd=2, col=2)
```

### Principal Component Analysis
```{r Genome: Principal Component Analysis}
PCs <- princomp(AIMs, scores=T)
PC.score <- PCs$scores
plot(PC.score[,1], PC.score[,2], col=(1+Pop), pch=16, xlab="PC1", ylab="PC2")

```

